### PR TITLE
Feature/pn 162 compose and handover sms to guble

### PR DESCRIPTION
**Pull request by marian-craciunescu**
WIP

### CODE REVIEW

File: server/gubled.go [hunk 1]
```diff
@@ -170,6 +171,25 @@ var CreateModules = func(router router.Router) []interface{} {
 		logger.Info("APNS: disabled")
 	}
 
+	if *Config.SMS.Enabled {
+		logger.Info("Optivo SMS: enabled")
```
Comment (Bogh) [inline server/gubled.go]: Remove Optivio from comment.

File: server/gubled.go [hunk 2]
```diff
@@ -170,6 +171,25 @@ var CreateModules = func(router router.Router) []interface{} {
...
+		if *Config.SMS.APIKey == "" || *Config.SMS.APISecret == "" {
+			logger.Panic("The API Key has to be provided when NEXMO SMS connector is enabled")
+		}
+		optivoSender, err := sms.NewNexmoSender(*Config.SMS.APIKey, *Config.SMS.APISecret)
+		if err != nil {
+			logger.WithError(err).Error("Error creating Optivo Sender")
```
Comment (Bogh) [inline server/gubled.go]: Remove  Optivio

File: server/gubled.go [hunk 3]
```diff
@@ -170,6 +171,25 @@ var CreateModules = func(router router.Router) []interface{} {
...
+		}
+		smsConn, err := sms.NewGateway(router, optivoSender, Config.SMS)
+		if err != nil {
+			logger.WithError(err).Error("Error creating Optivo Sender")
```
Comment (Bogh) [inline server/gubled.go]: Remove Optivio

File: server/gubled.go [hunk 4]
```diff
@@ -170,6 +171,25 @@ var CreateModules = func(router router.Router) []interface{} {
 		logger.Info("APNS: disabled")
 	}
 
+	if *Config.SMS.Enabled {
+		logger.Info("Optivo SMS: enabled")
+		if *Config.SMS.APIKey == "" || *Config.SMS.APISecret == "" {
+			logger.Panic("The API Key has to be provided when NEXMO SMS connector is enabled")
```
Comment (Bogh) [inline server/gubled.go]: Remove Nexmo from log message. We should make guble separated from provider as much as possible.

File: server/sms/gateway.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+	log "github.com/Sirupsen/logrus"
+	"github.com/smancke/guble/protocol"
+	"github.com/smancke/guble/server/connector"
+	"github.com/smancke/guble/server/router"
+	routerimport "github.com/smancke/guble/server/router"
+	"github.com/smancke/guble/server/store"
+	"strconv"
+)
+
+const SMS_SCHEMA = "sms_notifications"
```
Comment (Bogh) [inline server/sms/gateway.go]: Group these and use CamelCase since it's standard naming for constants.

File: server/sms/sms_test.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package gateway
+
+import (
+	"github.com/smancke/guble/testutil"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestNexmoSender_Send(t *testing.T) {
```
Comment (Bogh) [inline server/sms/sms_test.go]: This is an integration test? Will this send SMS directly to nexmo?
It should mock the nexmo api if possible.

File: server/sms/gateway.go [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+	logger *log.Entry
+	route  *router.Route
+
+	ctx        context.Context
+	cancel     context.CancelFunc
+	LastIDSent uint64
+}
+
+func NewGateway(router router.Router, sender Sender, config Config) (*gateway, error) {
+	if *config.Workers <= 0 {
```
Comment (Bogh) [inline server/sms/gateway.go]: Do we use workers in this connector?

File: server/sms/gateway.go [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+	APIKey    *string
+	APISecret *string
+	Workers   *int
+	SMSTopic  *string
+
+	Name      string
+	Schema    string
+}
+
+type gateway struct {
```
Comment (Bogh) [inline server/sms/gateway.go]: Would be ideal to name this Connector and the gateway to be the sender. The name gateway comes from SMS Gateway and it should be an interface that actually takes the destination and body and send it. And this library should be externally configured and not together with the SMS. (eg.: The api key and api secret is something specific to nexmo, may not be specific to a different provider)

File: server/sms/nexmo_sms_sender.go [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+	ApiKey    string `json:"api_key"`
+	ApiSecret string `json:"api_secret"`
+	To        string `json:"to"`
+	From      string `json:"from"`
+	SmsBody   string `json:"text"`
+}
+
+type ResponseCode int
+
+const (
```
Comment (Bogh) [inline server/sms/nexmo_sms_sender.go]: Please move constants definitions before type definitions.

File: server/sms/nexmo_sms_sender.go [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+	"strconv"
+)
+
+const (
+	KEY    = "ce40b46d"
+	SECRET = "153d2b2c72985370"
+	URL    = "https://rest.nexmo.com/sms/json?"
+)
+
+type NexmoSms struct {
```
Comment (Bogh) [inline server/sms/nexmo_sms_sender.go]: Keep type definition along side methods. You can even move this type in own file.

File: server/sms/sms_test.go (hunk 6 above)
Comment (marian-craciunescu) [inline server/sms/sms_test.go]: Yes this is  an integration test.l..What should I do with it ?

**Pull request merged**